File: MyRoomFE/MyRoomFE/Model/Used.swift
//
//  Used.swift
//  MyRoomFE
//
//  Created by 이수정 on 11/18/24.
//


import Foundation

struct Useds: Identifiable,Codable,Equatable {
    
    let id: Int
    let usedTitle: String
    let usedPrice: Int
    let usedDesc: String
    let user: User
    let usedUrl: String?
    let usedStatus: Int

    let usedPurchaseDate: String?
    let usedExpiryDate: String?
    let usedOpenDate: String?
    let purchasePrice: Int?

    let usedFavCnt: Int
    let usedViewCnt: Int
    let usedChatCnt: Int

    let images: [UsedPhotoData]
    let usedThumbnail: String
    var isFavorite:Bool
    let usedFav: [UsedFavData]?
    let updatedAt: String
    
    mutating func toggleFavorite() {
        isFavorite.toggle()
    }
    
    mutating func setFavorite(_ value: Bool) {
        isFavorite = value
    }
  
}

struct User : Codable,Equatable {
    let nickname:String
    let userImage:String?
}


struct UsedRoot: Codable{
//    let success: Bool
    let useds: [Useds]
//    let message: String
}




struct UsedPhotos: Codable,Equatable {
    let images: [UsedPhotoData]
}


struct UsedPhotoData:Identifiable,Codable, Equatable, Hashable  {
    let id:Int
    let image:String
}

struct UsedFavs:Codable,Equatable {
    let usedFav: [UsedFavData]
}

struct UsedFavData:Identifiable,Codable, Equatable {
    let id:Int
    let usedId:Int
    let userId:Int
}

File: MyRoomBE/models/used.js
'use strict';
const {
  Model
} = require('sequelize');
module.exports = (sequelize, DataTypes) => {
  class Used extends Model {
    /**
     * Helper method for defining associations.
     * This method is not a part of Sequelize lifecycle.
     * The `models/index` file will call this method automatically.
     */
    static associate(models) {
      Used.hasMany(models.UsedPhoto,{
        foreignKey:"usedId",
        as:"images"
      })

      Used.hasMany(models.UsedFav,{
        foreignKey:"usedId",
        as:"usedFav"
      })

      Used.belongsTo(models.User,{
        foreignKey:"userId",
        onDelete: "CASCADE",
        as:"user"
      })
      
    }
  }
  Used.init({
    usedTitle: DataTypes.STRING,
    usedPrice: DataTypes.INTEGER,
    usedDesc: DataTypes.STRING,
    usedPurchaseDate: DataTypes.DATE,
    usedExpiryDate: DataTypes.DATE,
    usedOpenDate: DataTypes.DATE,
    purchasePrice: DataTypes.INTEGER,
    usedViewCnt: DataTypes.INTEGER,
    usedChatCnt: DataTypes.INTEGER,
    usedFavCnt: DataTypes.INTEGER,
    usedStatus: DataTypes.INTEGER,
    userId: DataTypes.INTEGER,
    usedUrl:DataTypes.STRING,
    usedThumbnail: DataTypes.STRING,
    itemName: DataTypes.STRING,
    itemDesc: DataTypes.STRING,
  }, {
    sequelize,
    modelName: 'Used',
  });
  return Used;
};